File: packages/provide-uterm-go/shell/connector.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package shell

import (
	"context"
	"fmt"
	"hash/fnv"
	"sort"
	"strings"
	"sync"
	"time"
	"unicode/utf8"

	ptel "github.com/provide-io/provide-telemetry/go"
)

// defaultPollDelay is the idle poll back-off (matches _connector.py's 0.05s).
const defaultPollDelay = 50 * time.Millisecond

// ConnectorConfig configures a UshellConnector. It mirrors the Python
// constructor keyword arguments (display_name, _config, extra_ctx); _config is
// intentionally unused (reserved), as in the reference.
type ConnectorConfig struct {
	// DisplayName is a human-readable name (defaults to the session id).
	DisplayName string
	// ExtraCtx seeds the dispatcher context Values (Python extra_ctx). Ignored
	// when Context is set.
	ExtraCtx map[string]any
	// Context provides the typed CF bindings (Env / Storage / ListKVSessions).
	// When nil a context with Values=ExtraCtx is built.
	Context *Context
}

// UshellConnector is an interactive REPL connector that needs no external
// process. It provides the same method set as the Python UshellConnector
// (SessionConnector protocol): Start/Stop/IsConnected, PollMessages,
// HandleInput, HandleControl, GetSnapshot, GetAnalysis, Clear, SetMode.
//
// Unlike the single-threaded asyncio reference, this port is safe for
// concurrent use: all mutable state is guarded by mu, and animation streaming
// runs in a goroutine that appends frames under the same lock.
type UshellConnector struct {
	mu          sync.Mutex
	sessionID   string
	displayName string
	connected   bool
	welcomed    bool
	flowPaused  bool
	buf         *LineBuffer
	dispatcher  *CommandDispatcher
	ctxValues   map[string]any
	pending     []Frame
	animCancel  context.CancelFunc
	animDone    chan struct{}

	// pollSleep / pollDelay implement the idle back-off; both are overridable
	// in tests within the package.
	pollSleep func(time.Duration)
	pollDelay time.Duration
}

// NewUshellConnector builds a connector for the given session id.
func NewUshellConnector(sessionID string, cfg ConnectorConfig) *UshellConnector {
	dctx := cfg.Context
	if dctx == nil {
		dctx = &Context{Values: cfg.ExtraCtx}
	}
	display := cfg.DisplayName
	if display == "" {
		display = sessionID
	}
	return &UshellConnector{
		sessionID:   sessionID,
		displayName: display,
		buf:         NewLineBuffer(),
		dispatcher:  NewCommandDispatcher(dctx),
		ctxValues:   dctx.Values,
		pollSleep:   time.Sleep,
		pollDelay:   defaultPollDelay,
	}
}

// Start marks the connector connected.
func (c *UshellConnector) Start() {
	c.mu.Lock()
	c.connected = true
	c.mu.Unlock()
}

// Stop marks the connector disconnected and cancels any running animation.
func (c *UshellConnector) Stop() {
	c.mu.Lock()
	c.connected = false
	if c.animCancel != nil {
		c.animCancel()
	}
	c.mu.Unlock()
}

// IsConnected reports the connection state.
func (c *UshellConnector) IsConnected() bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.connected
}

// PollMessages returns the welcome frames on the first call after connect, then
// any pending frames, else nil after an idle back-off. Port of poll_messages.
func (c *UshellConnector) PollMessages() []Frame {
	c.mu.Lock()
	if !c.connected || c.flowPaused {
		c.mu.Unlock()
		return nil
	}
	if !c.welcomed {
		c.welcomed = true
		c.mu.Unlock()
		return []Frame{WorkerHello("open"), Term(Banner + Prompt)}
	}
	if len(c.pending) > 0 {
		frames := c.pending
		c.pending = nil
		c.mu.Unlock()
		return frames
	}
	c.mu.Unlock()
	c.pollSleep(c.pollDelay)
	return nil
}

// HandleInput processes raw keystroke data and returns terminal frames. Port of
// handle_input.
func (c *UshellConnector) HandleInput(ctx context.Context, data string) []Frame {
	c.mu.Lock()
	c.buf.Feed(data)
	echo := c.buf.TakeEcho()
	completed := c.buf.TakeCompleted()
	c.mu.Unlock()

	var frames []Frame
	if echo != "" {
		frames = append(frames, Term(echo))
	}

	logger := ptel.GetLogger(ctx, "provide.uterm.shell")
	for _, line := range completed {
		logger.Debug("dispatch", "line", line)
		result := c.dispatcher.Dispatch(ctx, line)
		if result.Animated != nil {
			c.startAnimation(*result.Animated)
			continue
		}
		for _, s := range result.Text {
			frames = append(frames, Term(s))
		}
	}
	return frames
}

// startAnimation cancels any running animation and launches a new streaming
// goroutine. Port of the AnimatedResult branch of handle_input +
// _stream_animation.
func (c *UshellConnector) startAnimation(result AnimatedResult) {
	c.mu.Lock()
	if c.animCancel != nil {
		c.animCancel()
	}
	actx, cancel := context.WithCancel(context.Background())
	done := make(chan struct{})
	c.animCancel = cancel
	c.animDone = done
	c.mu.Unlock()

	go c.streamAnimation(actx, done, result)
}

// streamAnimation streams frames with per-frame delay, appending them to the
// pending queue, then appends the prompt (on completion or cancellation).
func (c *UshellConnector) streamAnimation(ctx context.Context, done chan struct{}, result AnimatedResult) {
	defer close(done)
	delay := 100 * time.Millisecond
	if result.FPS > 0 {
		delay = time.Duration(float64(time.Second) / result.FPS)
	}
	for {
		for _, frame := range result.Frames {
			timer := time.NewTimer(delay)
			select {
			case <-ctx.Done():
				timer.Stop()
				c.appendPending(Term(Prompt))
				return
			case <-timer.C:
			}
			c.appendPending(Term(frame))
		}
		if !result.Loop {
			break
		}
		select {
		case <-ctx.Done():
			c.appendPending(Term(Prompt))
			return
		default:
		}
	}
	c.appendPending(Term(Prompt))
}

// waitAnimDone returns the current animation's completion channel, or nil when
// no animation has run. The channel is closed when the streaming goroutine
// exits.
func (c *UshellConnector) waitAnimDone() <-chan struct{} {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.animDone
}

// appendPending queues a frame for the next PollMessages.
func (c *UshellConnector) appendPending(f Frame) {
	c.mu.Lock()
	c.pending = append(c.pending, f)
	c.mu.Unlock()
}

// HandleControl handles flow_pause / flow_resume backpressure and
// snapshot_request; hijack actions (pause/resume/step) are no-ops. Port of
// handle_control.
func (c *UshellConnector) HandleControl(action string) []Frame {
	switch action {
	case "flow_pause":
		c.mu.Lock()
		c.flowPaused = true
		c.mu.Unlock()
	case "flow_resume":
		c.mu.Lock()
		c.flowPaused = false
		c.mu.Unlock()
	case "snapshot_request":
		return []Frame{c.GetSnapshot()}
	}
	return nil
}

// GetSnapshot returns a fresh screen snapshot frame. Port of get_snapshot.
func (c *UshellConnector) GetSnapshot() Frame {
	c.mu.Lock()
	current := c.buf.CurrentLine()
	sid := c.sessionID
	c.mu.Unlock()

	screen := "ushell " + sid + "\r\n" + Prompt + current
	x := utf8.RuneCountInString(Prompt) + utf8.RuneCountInString(current)
	return Frame{
		"type":               "snapshot",
		"screen":             screen,
		"cursor":             map[string]int{"x": x, "y": 1},
		"cols":               80,
		"rows":               24,
		"screen_hash":        screenHash(screen),
		"cursor_at_end":      true,
		"has_trailing_space": false,
		"prompt_detected":    map[string]string{"prompt_id": "ushell_prompt"},
		"ts":                 nowTS(),
	}
}

// GetAnalysis returns a human-readable analysis string. Port of get_analysis.
//
// Deviation: the Python "sandbox_names" line becomes "context_names" (the
// Python-exec sandbox is not ported), and booleans / reprs follow Go
// conventions rather than Python's.
func (c *UshellConnector) GetAnalysis() string {
	c.mu.Lock()
	connected := c.connected
	current := c.buf.CurrentLine()
	names := make([]string, 0, len(c.ctxValues))
	for k := range c.ctxValues {
		if !strings.HasPrefix(k, "__") {
			names = append(names, k)
		}
	}
	sid := c.sessionID
	c.mu.Unlock()
	sort.Strings(names)

	lines := []string{
		fmt.Sprintf("[ushell analysis — session: %s]", sid),
		fmt.Sprintf("connected: %t", connected),
		fmt.Sprintf("current_line: %q", current),
		fmt.Sprintf("context_names: %v", names),
	}
	return strings.Join(lines, "\n")
}

// Clear discards the current line and returns a clear-screen frame. Port of
// clear.
func (c *UshellConnector) Clear() []Frame {
	c.mu.Lock()
	c.buf.Clear()
	c.mu.Unlock()
	return []Frame{Term(ClearScreen + Prompt)}
}

// SetMode accepts a mode change; ushell is always open, so it just re-advertises
// the mode. Port of set_mode.
func (c *UshellConnector) SetMode(mode string) []Frame {
	return []Frame{WorkerHello(mode)}
}

// screenHash derives a short, deterministic hash of screen. Deviation: the
// Python str(hash(screen))[:16] uses Python's process-local, non-portable
// hash; this uses FNV-1a. Both are session-local snapshot hints, not wire
// contracts.
func screenHash(screen string) string {
	h := fnv.New64a()
	_, _ = h.Write([]byte(screen))
	s := fmt.Sprintf("%d", h.Sum64())
	if len(s) > 16 {
		s = s[:16]
	}
	return s
}

File: packages/provide-uterm-csharp/src/Provide.Uterm/Shell/Connector.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

using System.Security.Cryptography;
using System.Text;
using Provide.Uterm.Connectors;
using Provide.Uterm.Session;
using Provide.Uterm.TermSession;

namespace Provide.Uterm.Shell;

/// <summary>Config for <see cref="UshellConnector"/>. Port of Go shell.ConnectorConfig.</summary>
public sealed class UshellConnectorConfig
{
    public string DisplayName { get; init; } = "";
    public ShellContext? Context { get; init; }
    public Dictionary<string, object?>? ExtraCtx { get; init; }
    public HttpClient? HttpClient { get; init; }
    public Func<byte[], int, int, string, (IReadOnlyList<string> Frames, double Fps)>? RenderImage { get; init; }

    /// <summary>Test hook: replace idle poll sleep (Go pollSleep).</summary>
    public Action<TimeSpan>? PollSleep { get; init; }

    public TimeSpan PollDelay { get; init; } = TimeSpan.FromMilliseconds(50);
}

/// <summary>
/// Interactive REPL connector (no external process). Port of Go shell.UshellConnector
/// and Python UshellConnector: Start/Stop/IsConnected, PollMessages, HandleInput,
/// HandleControl, GetSnapshot, GetAnalysis, Clear, SetMode.
/// Also implements <see cref="IConnector"/> for registry/session wiring.
/// </summary>
public sealed class UshellConnector : IConnector
{
    private readonly object _gate = new();
    private readonly string _sessionId;
    private readonly string _displayName;
    private readonly LineBuffer _buf = new();
    private readonly CommandDispatcher _dispatcher;
    private readonly Dictionary<string, object?> _ctxValues;
    private readonly Action<TimeSpan> _pollSleep;
    private readonly TimeSpan _pollDelay;
    private readonly List<Dictionary<string, object?>> _pending = new();

    private bool _connected;
    private bool _welcomed;
    private bool _flowPaused;
    private CancellationTokenSource? _animCts;
    private Task? _animTask;

    public UshellConnector(string sessionId, UshellConnectorConfig? config = null)
    {
        config ??= new UshellConnectorConfig();
        _sessionId = sessionId;
        _displayName = string.IsNullOrEmpty(config.DisplayName) ? sessionId : config.DisplayName;
        var ctx = config.Context ?? new ShellContext
        {
            Values = config.ExtraCtx is null
                ? new Dictionary<string, object?>()
                : new Dictionary<string, object?>(config.ExtraCtx),
        };
        _ctxValues = ctx.Values;
        _dispatcher = new CommandDispatcher(ctx, config.HttpClient, config.RenderImage);
        _pollSleep = config.PollSleep ?? (d =>
        {
            if (d > TimeSpan.Zero)
            {
                Thread.Sleep(d);
            }
        });
        _pollDelay = config.PollDelay <= TimeSpan.Zero ? TimeSpan.FromMilliseconds(50) : config.PollDelay;
    }

    public string SessionId => _sessionId;
    public string DisplayName => _displayName;

    public void Start()
    {
        lock (_gate)
        {
            _connected = true;
        }
    }

    public void Stop()
    {
        CancellationTokenSource? cts;
        lock (_gate)
        {
            _connected = false;
            cts = _animCts;
            _animCts = null;
        }

        cts?.Cancel();
        cts?.Dispose();
    }

    public bool IsConnected()
    {
        lock (_gate)
        {
            return _connected;
        }
    }

    /// <summary>Welcome frames then pending; idle poll sleeps then returns empty.</summary>
    public IReadOnlyList<Dictionary<string, object?>> PollMessages()
    {
        lock (_gate)
        {
            if (!_connected || _flowPaused)
            {
                return Array.Empty<Dictionary<string, object?>>();
            }

            if (!_welcomed)
            {
                _welcomed = true;
                return new[]
                {
                    ShellFrames.WorkerHello("open"),
                    ShellFrames.Term(ShellOutput.Banner + ShellOutput.Prompt),
                };
            }

            if (_pending.Count > 0)
            {
                var frames = _pending.ToList();
                _pending.Clear();
                return frames;
            }
        }

        _pollSleep(_pollDelay);
        return Array.Empty<Dictionary<string, object?>>();
    }

    public IReadOnlyList<Dictionary<string, object?>> HandleInput(string data)
    {
        string echo;
        IReadOnlyList<string> completed;
        lock (_gate)
        {
            _buf.Feed(data);
            echo = _buf.TakeEcho();
            completed = _buf.TakeCompleted();
        }

        var frames = new List<Dictionary<string, object?>>();
        if (echo.Length > 0)
        {
            frames.Add(ShellFrames.Term(echo));
        }

        foreach (var line in completed)
        {
            var result = _dispatcher.Dispatch(line);
            if (result.Animated is not null)
            {
                StartAnimation(result.Animated);
                continue;
            }

            foreach (var s in result.Text)
            {
                frames.Add(ShellFrames.Term(s));
            }
        }

        return frames;
    }

    public IReadOnlyList<Dictionary<string, object?>> HandleControl(string action)
    {
        switch (action)
        {
            case "flow_pause":
                lock (_gate)
                {
                    _flowPaused = true;
                }

                return Array.Empty<Dictionary<string, object?>>();
            case "flow_resume":
                lock (_gate)
                {
                    _flowPaused = false;
                }

                return Array.Empty<Dictionary<string, object?>>();
            case "snapshot_request":
                return new[] { GetSnapshot() };
            default:
                return Array.Empty<Dictionary<string, object?>>();
        }
    }

    public Dictionary<string, object?> GetSnapshot()
    {
        string current;
        string sid;
        lock (_gate)
        {
            current = _buf.CurrentLine();
            sid = _sessionId;
        }

        var screen = "ushell " + sid + "\r\n" + ShellOutput.Prompt + current;
        var x = ShellOutput.Prompt.Length + current.Length; // ASCII-aligned like Go rune count for ASCII prompt
        return new Dictionary<string, object?>
        {
            ["type"] = "snapshot",
            ["screen"] = screen,
            ["cursor"] = new Dictionary<string, object?> { ["x"] = x, ["y"] = 1 },
            ["cols"] = 80,
            ["rows"] = 24,
            ["screen_hash"] = ScreenHash(screen),
            ["cursor_at_end"] = true,
            ["has_trailing_space"] = false,
            ["prompt_detected"] = new Dictionary<string, object?> { ["prompt_id"] = "ushell_prompt" },
            ["ts"] = ShellFrames.NowTs(),
        };
    }

    public string GetAnalysis()
    {
        bool connected;
        string current;
        string sid;
        List<string> names;
        lock (_gate)
        {
            connected = _connected;
            current = _buf.CurrentLine();
            sid = _sessionId;
            names = _ctxValues.Keys.Where(k => !k.StartsWith("__", StringComparison.Ordinal)).OrderBy(k => k, StringComparer.Ordinal).ToList();
        }

        return string.Join(
            "\n",
            $"[ushell analysis — session: {sid}]",
            $"connected: {connected.ToString().ToLowerInvariant()}",
            $"current_line: \"{current}\"",
            $"context_names: [{string.Join(' ', names)}]");
    }

    public IReadOnlyList<Dictionary<string, object?>> ClearScreen()
    {
        lock (_gate)
        {
            _buf.Clear();
        }

        return new[] { ShellFrames.Term(ShellOutput.ClearScreen + ShellOutput.Prompt) };
    }

    public IReadOnlyList<Dictionary<string, object?>> SetMode(string mode) =>
        new[] { ShellFrames.WorkerHello(mode) };

    // --- IConnector surface (session registry / hosted runtime) ---

    public Task StartAsync(CancellationToken cancellationToken = default)
    {
        cancellationToken.ThrowIfCancellationRequested();
        Start();
        return Task.CompletedTask;
    }

    public Task StopAsync(CancellationToken cancellationToken = default)
    {
        cancellationToken.ThrowIfCancellationRequested();
        Stop();
        return Task.CompletedTask;
    }

    public Task HandleInputAsync(string data, CancellationToken cancellationToken = default)
    {
        cancellationToken.ThrowIfCancellationRequested();
        var frames = HandleInput(data);
        lock (_gate)
        {
            _pending.AddRange(frames);
        }

        return Task.CompletedTask;
    }

    void IConnector.HandleControl(string action) => _ = HandleControl(action);

    void IConnector.SetMode(string mode)
    {
        if (mode is not ("open" or "hijack"))
        {
            throw new ArgumentException($"invalid mode: {mode}");
        }

        lock (_gate)
        {
            _pending.AddRange(SetMode(mode));
        }
    }

    void IConnector.Clear()
    {
        lock (_gate)
        {
            _pending.AddRange(ClearScreen());
        }
    }

    Snapshot IConnector.Snapshot()
    {
        var f = GetSnapshot();
        return new Snapshot
        {
            Screen = f.TryGetValue("screen", out var s) ? s?.ToString() ?? "" : "",
            Cols = 80,
            Rows = 24,
        };
    }

    string IConnector.Analysis() => GetAnalysis();

    IReadOnlyList<Dictionary<string, object?>> IConnector.Events()
    {
        // Hosted runtime drains frames via Events(); reuse PollMessages without idle sleep.
        lock (_gate)
        {
            if (!_connected || _flowPaused)
            {
                return Array.Empty<Dictionary<string, object?>>();
            }

            if (!_welcomed)
            {
                _welcomed = true;
                return new List<Dictionary<string, object?>>
                {
                    ShellFrames.WorkerHello("open"),
                    ShellFrames.Term(ShellOutput.Banner + ShellOutput.Prompt),
                };
            }

            if (_pending.Count == 0)
            {
                return Array.Empty<Dictionary<string, object?>>();
            }

            var copy = _pending.ToList();
            _pending.Clear();
            return copy;
        }
    }

    TransportSession? IConnector.Session() => null;

    private void StartAnimation(AnimatedResult result)
    {
        CancellationTokenSource? old;
        lock (_gate)
        {
            old = _animCts;
            _animCts = new CancellationTokenSource();
            var ct = _animCts.Token;
            _animTask = Task.Run(() => StreamAnimationAsync(result, ct), CancellationToken.None);
        }

        try
        {
            old?.Cancel();
        }
        catch
        {
            // ignore
        }

        old?.Dispose();
    }

    private async Task StreamAnimationAsync(AnimatedResult result, CancellationToken ct)
    {
        var delay = result.Fps > 0
            ? TimeSpan.FromSeconds(1.0 / result.Fps)
            : TimeSpan.FromMilliseconds(100);
        try
        {
            while (true)
            {
                foreach (var frame in result.Frames)
                {
                    await Task.Delay(delay, ct).ConfigureAwait(false);
                    AppendPending(ShellFrames.Term(frame));
                }

                if (!result.Loop)
                {
                    break;
                }
            }

            AppendPending(ShellFrames.Term(ShellOutput.Prompt));
        }
        catch (OperationCanceledException)
        {
            AppendPending(ShellFrames.Term(ShellOutput.Prompt));
        }
    }

    private void AppendPending(Dictionary<string, object?> frame)
    {
        lock (_gate)
        {
            _pending.Add(frame);
        }
    }

    private static string ScreenHash(string screen)
    {
        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(screen));
        // Deterministic short decimal-like hint (not FNV but stable across process)
        var n = BitConverter.ToUInt64(hash, 0);
        var s = n.ToString(System.Globalization.CultureInfo.InvariantCulture);
        return s.Length > 16 ? s[..16] : s;
    }
}
